File: NetFramework/Ribbon/CS/ControlHost/ControlHostLib/PictureBoxHost.cs
using System;
using System.ComponentModel;
using System.Drawing;
using System.Windows.Forms;
using System.IO;
using System.Runtime.Serialization.Formatters.Binary;
using System.Xml;

using C1.Win.C1Ribbon;

namespace ControlHostLib
{
    /// <summary>
    /// The Ribbon element that contains a PictureBox.
    /// </summary>
    public class PictureBoxHost : RibbonControlHost
    {
        public const string XmlNodeName = "pictureBoxHost";

        /// <summary>
        /// Creates the PictureBox control and its host element.
        /// </summary>
        public PictureBoxHost()
            : base(new PictureBox())
        {
        }

        /// <inheritdoc />
        public override string ControlHostNodeName
        {
            get { return XmlNodeName; }
        }

        /// <summary>
        /// Gets the hosted PictureBox control.
        /// </summary>
        [Browsable(false)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public PictureBox PictureBox
        {
            get { return (PictureBox)Control; }
        }

        /// <summary>
        /// Gets or sets the image that the PictureBox displays.
        /// </summary>
        [Category("Appearance")]
        [Description("Gets or sets the image that the PictureBox displays.")]
        public Image Image
        {
            get { return PictureBox.Image; }
            set { PictureBox.Image = value; }
        }

        /// <inheritdoc />
        [Browsable(false)]
        [DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
        public override string Text
        {
            get { return string.Empty; }
            set { }
        }

        protected override void LoadControlHostProperties(XmlNode node)
        {
            base.LoadControlHostProperties(node);
            Image = ReadImage(node, "image");
        }

        protected override void SaveControlHostProperties(XmlWriter writer)
        {
            base.SaveControlHostProperties(writer);
            if (Image != null)
            {
                WriteImage(writer, "image", Image);
            }
        }

        static Image ReadImage(XmlNode node, string prefix)
        {
            XmlAttribute att = node.Attributes[prefix + "Base64"];
            if (att != null)
            {
                byte[] buf = Convert.FromBase64String(att.Value);
                MemoryStream ms = new MemoryStream(buf);
                BinaryFormatter bf = new BinaryFormatter();
                return bf.Deserialize(ms) as Image;
            }
            return null;
        }

        static void WriteImage(XmlWriter writer, string prefix, Image image)
        {
            if (image != null)
            {
                MemoryStream ms = new MemoryStream();
                BinaryFormatter bf = new BinaryFormatter();
                bf.Serialize(ms, image);
                string base64 = Convert.ToBase64String(ms.GetBuffer(), 0, (int)ms.Length);
                writer.WriteAttributeString(prefix + "Base64", base64);
            }
        }
    }
}

File: ExcludedSamples/Ribbon/VB/ControlHost/ControlHostLib/PictureBoxHost.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.IO
Imports System.Runtime.Serialization.Formatters.Binary
Imports System.Windows.Forms
Imports System.Xml
Imports C1.Win.C1Ribbon

''' <summary>
''' The Ribbon element that contains a PictureBox.
''' </summary>
Public Class PictureBoxHost
    Inherits RibbonControlHost

    Public Const XmlNodeName As String = "pictureBoxHost"

    ''' <summary>
    ''' Creates the PictureBox control and its host element.
    ''' </summary>
    Public Sub New()
        MyBase.New(New PictureBox())
    End Sub

    ''' <summary>
    ''' <inheritdoc />
    ''' </summary>
    Public Overrides ReadOnly Property ControlHostNodeName As String
        Get
            Return XmlNodeName
        End Get
    End Property

    ''' <summary>
    ''' Gets the hosted PictureBox control.
    ''' </summary>
    <Browsable(False)> _
    <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
    Public ReadOnly Property PictureBox() As PictureBox
        Get
            Return CType(Control, PictureBox)
        End Get
    End Property

    ''' <summary>
    ''' Gets or sets the image that the PictureBox displays.
    ''' </summary>
    <Category("Appearance")> _
    <Description("Gets or sets the image that the PictureBox displays.")> _
    Public Property Image() As Image
        Get
            Return PictureBox.Image
        End Get
        Set(ByVal value As Image)
            PictureBox.Image = value
        End Set
    End Property

    ''' <summary>
    ''' This property is not applicable to the PictureBox.
    ''' </summary>
    <Browsable(False)> _
    <DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)> _
    Public Overrides Property Text() As String
        Get
            Return MyBase.Text
        End Get
        Set(ByVal value As String)
            MyBase.Text = value
        End Set
    End Property

    Protected Overrides Sub LoadControlHostProperties(ByVal node As XmlNode)
        MyBase.LoadControlHostProperties(node)
        Image = ReadImage(node, "image")
    End Sub

    Protected Overrides Sub SaveControlHostProperties(ByVal writer As XmlWriter)
        MyBase.SaveControlHostProperties(writer)
        If Image IsNot Nothing Then
            WriteImage(writer, "image", Image)
        End If
    End Sub

    Private Shared Function ReadImage(ByVal node As XmlNode, ByVal prefix As String) As Image
        Dim att As XmlAttribute = node.Attributes(prefix & "Base64")
        If att IsNot Nothing Then
            Dim buf As Byte() = Convert.FromBase64String(att.Value)
            Dim ms As MemoryStream = New MemoryStream(buf)
            Dim bf As BinaryFormatter = New BinaryFormatter()
            Return TryCast(bf.Deserialize(ms), Image)
        End If

        Return Nothing
    End Function

    Private Shared Sub WriteImage(ByVal writer As XmlWriter, ByVal prefix As String, ByVal image As Image)
        If image IsNot Nothing Then
            Dim ms As MemoryStream = New MemoryStream()
            Dim bf As BinaryFormatter = New BinaryFormatter()
            bf.Serialize(ms, image)
            Dim base64 As String = Convert.ToBase64String(ms.GetBuffer(), 0, CInt(ms.Length))
            writer.WriteAttributeString(prefix & "Base64", base64)
        End If
    End Sub

End Class
